myhpi/tests/polls/test_ranked_choice_algorithm.py

```diff
@@ -0,0 +1,107 @@
+        tags = getattr(method, "tags", {})
+        if "skip_setup" in tags:
+            return
+        self.setup_poll()
+
+        self.options = {
+            "alice": RankedChoiceOption.objects.create(name="Alice", poll=self.poll),
+            "bob": RankedChoiceOption.objects.create(name="Bob", poll=self.poll),
+            "charlie": RankedChoiceOption.objects.create(name="Charlie", poll=self.poll),
+            "dora": RankedChoiceOption.objects.create(name="Dora", poll=self.poll),
+        }
+
+    def setup_poll(self):
+        self.poll_list = PollList(
+            title="Polls",
+            slug="polls",
+            path="0001000200010005",
+            depth=4,
+            is_public=True,
+        )
+        self.information_menu.add_child(instance=self.poll_list)
+
+        self.poll = RankedChoicePoll(
+            title="How are you?",
+            slug="how-are-you",
+            description="This is a poll to check how you are.",
+            start_date=datetime.now() - timedelta(days=1),
+            end_date=datetime.now() + timedelta(days=1),
+            results_visible=False,
+            is_public=True,
+        )
+
+        self.poll_list.add_child(instance=self.poll)
+
+    def cast_ballots(self, template_ballots):
+        for template_ballot in template_ballots:
+            ballot = RankedChoiceBallot.objects.create(poll=self.poll)
+            ballot.save()
+
+            for idx, template_entry in enumerate(template_ballot):
+                RankedChoiceBallotEntry.objects.create(
+                    ballot=ballot, option=self.options[template_entry], rank=idx
+                )
+
+    def test_can_two_first_places(self):
+        self.cast_ballots([["alice", "bob"], ["bob", "alice"]])
+        self.assertEqual(
+            self.poll.calculate_ranking(),
+            [(1, "Alice", 1), (1, "Bob", 1), (3, "Charlie", 0), (3, "Dora", 0)],
+        )
+
+    def test_three_first_places(self):
+        self.cast_ballots([["alice", "bob"], ["bob", "alice"], ["charlie"]])
+        self.assertEqual(
+            self.poll.calculate_ranking(),
+            [(1, "Alice", 1), (1, "Bob", 1), (1, "Charlie", 1), (4, "Dora", 0)],
+        )
+
+    def test_runoff_with_skips(self):
+        self.cast_ballots(
+            ([["alice", "charlie", "bob"]] * 10)
+            + ([["bob", "alice", "charlie"]] * 12)
+            + ([["charlie", "dora", "alice", "bob"]] * 9)
+            + ([["dora", "alice", "charlie"]] * 1)
+        )
+        self.assertEqual(
+            self.poll.calculate_ranking(),
+            [(1, "Alice", 32), (2, "Bob", 12), (3, "Charlie", 9), (4, "Dora", 1)],
+        )
+
+    def test_no_ballots(self):
+        self.assertEqual(
+            self.poll.calculate_ranking(),
+            [(1, "Alice", 0), (1, "Bob", 0), (1, "Charlie", 0), (1, "Dora", 0)],
+        )
+
+    def test_fast(self):
+        # TODO: Please assert this is reasonably fast. Do your Django magic here.
```

Comment:
Who is this addressed to?

Reply:
I already asked both DuckDuckGo and @jeriox, but got no better answer then just timing it, so I just implemented that in 9a546d5024c470c177e6fb8dca354a3fe4dff3d3

Though, if you have a better idea, I'd also try that.